docs/proposals/update-machine.md

```diff
@@ -0,0 +1,163 @@
+shortretry-->end1
+
+begin-->AddBootstrapTokenToUserData
+-->gms["statusResp,statusErr=driver.GetMachineStatus(...)"]
+-->chkstatuserr{"Checkk statusErr"}
+chkstatuserr--notFound-->chknodelbl{"Chk Node Label"}
+chkstatuserr--else-->createFailed["lastop.Type=Create,lastop.state=Failed,phase=CrashLoopBackOff|Failed"]-->medretry
+chkstatuserr--nil-->initnodename["nodeName = statusResp.NodeName"]-->setnodename
+
+
+chknodelbl--notset-->createmachine["createResp, createErr=driver.CreateMachine(...)"]-->chkCreateErr{"Check createErr"}
+
+chkCreateErr--notnil-->createFailed
+
+chkCreateErr--nil-->getnodename["nodeName = createResp.NodeName"]
+-->chkstalenode{"nodeName != machine.Name\n//chk stale node"}
+chkstalenode--false-->setnodename["if unset machine.Labels['node']= nodeName"]
+-->machinepending["if empty/crashloopbackoff lastop.type=Create,lastop.State=Processing,phase=Pending"]
+-->shortretry
+
+chkstalenode--true-->delmachine["driver.DeleteMachine(...)"]
+-->permafail["lastop.type=Create,lastop.state=Failed,Phase=Failed"]
+-->shortretry
+
+subgraph noteA [" "]
+    permafail -.- note1(["VM was referring to stale node obj"])
+end
+style noteA opacity:0
+
+
+subgraph noteB [" "]
+    createFailed-.- note2(["Proposal: Enhance handling to support startup logic"])
+end
+```
+
+### Enhancement of MCM triggerCreationFlow
+
+The following changes are proposed with a view towards minimal impact on current code and no introduction of a new Machine Phase.
+
+1. We need to introduce enhancements around the call to `Driver.CreateMachine` in the flow illustrated above.
+1. Observe that after the call to a successful `Driver.CreateMachine`, the machine phase is set to `Pending`, the `LastOperation.Type` is currently set to `Create` and the `LastOperation.State` set to `Processing`before returning with a `ShortRetry`. The `LastOperation.Description` is (unfortunately) set to the fixed message: `Creating machine on cloud provider`.
+1. Observe that after an erroneous call to `Driver.CreateMachine`, the machine phase is set to `CrashLoopBackOff` or `Failed`  (in case of creation timeout).
+1. We propose introducing a new MC error code `codes.Startup` indicating that the VM Instance was created but there was a startup error in post VM creation activities. The implementor of `Driver.CreateMachine` can return this error code, indicating that `CreateMachine` needs to be called again. However, the Machine Controller will keep the machine phase for such a situation as  `Pending` and not change the phase to `CrashLoopBackOff` when encountering a `codes.Startup` error. (The instance for the machine HAS been created hence `CrashLoopBackoff` is inappropriate)
+1. We will introduce a new _machine operation_ stage `InitiateStartup`. In case of an `codes.Startup` error, the `machine.Status.LastOperation.Description` will be set to this stage, the `LastOperation.Type` will be set to `Create` and the `LastOperation.State` set to `Failed` before returning with a `ShortRetry`
```

Comment:
There are situations which we should think of dealing with. For example, what if following condition happens
- VM was created && 
- there was a startup failure &&,
-  `status.LastOperation` couldn't be updated due to KAPI rate limiting or MCM restart 

then in the next reconcile, we wouldn't know at what step we are

`GetMachineStatus` is used today to see if VM exists ,but it won't be able to help in identifying if startup failure happened or not.
This is happening as the `CreateMachine` call won't be atomic as per the proposal (its not atomic in Azure as well presently, but there we separately list nic to see where we have reached in create operation)

Probable solutions to the above situation:
* always assume StartupFailure happened in such a situation, let `CreateMachine()` be called , but enhance `CreateMachine()` to first check if the VM has all the startup settings done(which I'm not sure is possible for every startup setting), if not then call startup related calls

Reply:
Ideally `GetMachineStatus` should be enhanced to give the status of the machine keeping in mind the desired state of the machine vs as-is. This means that in cases where a `Create` involves creating the VM + updating properties on the VM (e.g. ipv6 specific configuration or source-destination-checks) then that should be separately sent back as part of the response. 

MCM can then decide what call to make based on the status returned.

Reply:
Agree it the clearly the job of `GetMachineStatus` to report this correctly. Provider implementations can simply return the same error proposed error code when the VM exists but startup not fully done.  `codes.Startup(Error)` (Himanshu preference) or `codes.PostCreate(Error)`. (Madhav preference)
 
 Please note that suffixing `Error` to the error codes currently breaks the convention in `codes`. There is no `Error` suffix in any of the error codes and I would like to preserve existing convention.

Reply:
Addressed point of modification to `GetMachineStatus` semantics

Reply:
Guys, after thinking on this a lot today and yesterday and realizing that provider implementations are now forced to check whether there is a VM exists but not is fully initialized and that this logic is now forced into BOTH `GetMachineStatus` and `CreateMachine`, I am proposing that we stick to a new method `InitializeMachine` which is distinct from `UpdateMachine` and is meant for instance initialization. This makes logic must simpler since `CreateMachine` complexity is kept minimal.

Reply:
Due to the point that Himanshu, mentioned above, the job of `Driver.CreateMachine` is becoming more and more complex. After thinking on this yesterday and today, I would propose simplifying this by introducing a new `Driver.InitializeMachine` which is separate from `Driver.UpdateMachine`. All post-create initialization logic set such as disabling src-dest checks (aws) and ipv6 address assignment (aws), etc. will be put in here. This method will be invoked only once after `Driver.CreateMachine` is invoked successfully. If `Driver.InitializeMachine` fails it will return `status.Error(codes.Initialization)` as the error. `Driver.GetMachine` also will return `status.Error(codes.Initialization)` if the VM is not fully initialized. The MCM will invoke `Driver.InitializeMachine` if `Driver.GetMachine` returns such a status.